// @electron-forge/maker-deb/dist/MakerDeb.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "MakerDebConfig", {
    enumerable: true,
    get: function() {
        return _config.MakerDebConfig;
    }
});
exports.debianArch = debianArch;
exports.default = void 0;
var _makerBase = _interopRequireDefault(require("@electron-forge/maker-base"));
var _path = _interopRequireDefault(require("path"));
var _config = require("./Config");
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function debianArch(nodeArch) {
    switch(nodeArch){
        case 'ia32':
            return 'i386';
        case 'x64':
            return 'amd64';
        case 'armv7l':
            return 'armhf';
        case 'arm':
            return 'armel';
        default:
            return nodeArch;
    }
}
class MakerDeb extends _makerBase.default {
    isSupportedOnCurrentPlatform() {
        return this.isInstalled('electron-installer-debian');
    }
    async make({ dir , makeDir , targetArch  }) {
        // eslint-disable-next-line global-require, import/no-unresolved, node/no-missing-require
        const installer = require('electron-installer-debian');
        const outDir = _path.default.resolve(makeDir, 'deb', targetArch);
        await this.ensureDirectory(outDir);
        const { packagePaths  } = await installer({
            options: {},
            ...this.config,
            arch: debianArch(targetArch),
            src: dir,
            dest: outDir,
            rename: undefined
        });
        return packagePaths;
    }
    constructor(...args){
        super(...args);
        this.name = 'deb';
        this.defaultPlatforms = [
            'linux'
        ];
        this.requiredExternalBinaries = [
            'dpkg',
            'fakeroot'
        ];
    }
}
exports.default = MakerDeb;
