// @electron-forge/installer-zip/dist/InstallerZip.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "InstallerOptions", {
    enumerable: true,
    get: function() {
        return _installerDarwin.InstallerOptions;
    }
});
exports.default = void 0;
var _installerDarwin = _interopRequireWildcard(require("@electron-forge/installer-darwin"));
var _crossSpawnPromise = require("@malept/cross-spawn-promise");
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
        return obj;
    } else {
        var newObj = {};
        if (obj != null) {
            for(var key in obj){
                if (Object.prototype.hasOwnProperty.call(obj, key)) {
                    var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};
                    if (desc.get || desc.set) {
                        Object.defineProperty(newObj, key, desc);
                    } else {
                        newObj[key] = obj[key];
                    }
                }
            }
        }
        newObj.default = obj;
        return newObj;
    }
}
class InstallerZip extends _installerDarwin.default {
    async install({ filePath , installSpinner  }) {
        await (0, _crossSpawnPromise).spawn('unzip', [
            '-q',
            '-o',
            _path.default.basename(filePath)
        ], {
            cwd: _path.default.dirname(filePath)
        });
        const appPath = (await _fsExtra.default.readdir(_path.default.dirname(filePath))).filter((file)=>file.endsWith('.app')
        ).map((file)=>_path.default.resolve(_path.default.dirname(filePath), file)
        ).sort((fA, fB)=>_fsExtra.default.statSync(fA).ctime.getTime() - _fsExtra.default.statSync(fB).ctime.getTime()
        )[0];
        const targetApplicationPath = `/Applications/${_path.default.basename(appPath)}`;
        await this.moveApp(appPath, targetApplicationPath, installSpinner);
        await (0, _crossSpawnPromise).spawn('open', [
            '-R',
            targetApplicationPath
        ], {
            detached: true
        });
    }
    constructor(...args){
        super(...args);
        this.name = 'zip';
    }
}
exports.default = InstallerZip;
