// @electron-forge/maker-dmg/dist/MakerDMG.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "MakerDMGConfig", {
    enumerable: true,
    get: function() {
        return _config.MakerDMGConfig;
    }
});
exports.default = void 0;
var _makerBase = _interopRequireDefault(require("@electron-forge/maker-base"));
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
var _config = require("./Config");
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
class MakerDMG extends _makerBase.default {
    isSupportedOnCurrentPlatform() {
        return process.platform === 'darwin';
    }
    async make({ dir , makeDir , appName , packageJSON , targetArch  }) {
        // eslint-disable-next-line global-require
        const electronDMG = require('electron-installer-dmg');
        const outPath = _path.default.resolve(makeDir, `${this.config.name || appName}.dmg`);
        const forgeDefaultOutPath = _path.default.resolve(makeDir, `${appName}-${packageJSON.version}-${targetArch}.dmg`);
        await this.ensureFile(outPath);
        const dmgConfig = {
            overwrite: true,
            name: appName,
            ...this.config,
            appPath: _path.default.resolve(dir, `${appName}.app`),
            out: _path.default.dirname(outPath)
        };
        const opts = await electronDMG(dmgConfig);
        if (!this.config.name) {
            await this.ensureFile(forgeDefaultOutPath);
            await _fsExtra.default.rename(outPath, forgeDefaultOutPath);
            return [
                forgeDefaultOutPath
            ];
        }
        return [
            opts.dmgPath
        ];
    }
    constructor(...args){
        super(...args);
        this.name = 'dmg';
        this.defaultPlatforms = [
            'darwin',
            'mas'
        ];
    }
}
exports.default = MakerDMG;
