// sponsorblock-api/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseError = exports.SponsorBlockAdmin = exports.SponsorBlockVIP = exports.SponsorBlock = exports.defaultOptions = void 0;
exports.defaultOptions = {
    baseURL: 'https://sponsor.ajay.app',
    hashPrefixLength: 4,
    service: 'YouTube',
    userAgent: `node-sponsorblock`,
};
var user_1 = require("./api/impl/user");
Object.defineProperty(exports, "SponsorBlock", { enumerable: true, get: function () { return user_1.SponsorBlock; } });
var vip_1 = require("./api/impl/vip");
Object.defineProperty(exports, "SponsorBlockVIP", { enumerable: true, get: function () { return vip_1.SponsorBlockVIP; } });
var admin_1 = require("./api/impl/admin");
Object.defineProperty(exports, "SponsorBlockAdmin", { enumerable: true, get: function () { return admin_1.SponsorBlockAdmin; } });
var ResponseError_1 = require("./errors/ResponseError");
Object.defineProperty(exports, "ResponseError", { enumerable: true, get: function () { return ResponseError_1.ResponseError; } });
__exportStar(require("./utils"), exports);
__exportStar(require("./types"), exports);
