// @electron-forge/async-ora/dist/ora-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _chalk = _interopRequireDefault(require("chalk"));
var _ora = _interopRequireDefault(require("./ora"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
class OraImpl {
    // eslint-disable-next-line no-empty-function, no-useless-constructor
    constructor(text = ''){
        this.text = text;
    }
    succeed(_symbol) {
        return this;
    }
    fail(_symbol) {
        return this;
    }
    start() {
        return this;
    }
    stop(_symbol) {
        return this;
    }
    warn(_message) {
        return this;
    }
}
exports.OraImpl = OraImpl;
const asyncOra = (initialOraValue, asyncFn, processExitFn = process.exit)=>{
    let fnOra = new OraImpl(initialOraValue);
    if (asyncOra.interactive) {
        fnOra = (0, _ora).default(initialOraValue).start();
    }
    return new Promise((resolve, reject)=>{
        asyncFn(fnOra).then(()=>{
            const wasPaused = process.stdin.isPaused();
            // Note: this may pause stdin as a side-effect in certain cases
            fnOra.succeed();
            if (asyncOra.keepStdinFlowing && !wasPaused && process.stdin.isPaused()) {
                process.stdin.resume();
            }
            return resolve();
        }).catch((err)=>{
            fnOra.fail();
            if (asyncOra.interactive) {
                if (err && err.message && err.stack) {
                    console.error(_chalk.default.red('\nAn unhandled error has occurred inside Forge:'));
                    console.error(_chalk.default.red(err.message));
                    console.error(_chalk.default.red(err.stack));
                } else {
                    console.error(_chalk.default.red('\nElectron Forge was terminated:'));
                    console.error(_chalk.default.red(typeof err === 'string' ? err : JSON.stringify(err)));
                }
                processExitFn(1);
                // If the process is still alive we should continue because either
                // something went really wrong or we are testing this function
                setTimeout(()=>resolve()
                , 500);
            } else {
                reject(err);
            }
        });
    });
};
asyncOra.interactive = true;
var _default = asyncOra;
exports.default = _default;
