// @electron-forge/async-ora/dist/ora.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.fakeOra = void 0;
var _chalk = _interopRequireDefault(require("chalk"));
var _debug = _interopRequireDefault(require("debug"));
require("log-symbols");
var _ora = _interopRequireDefault(require("ora"));
var _prettyMs = _interopRequireDefault(require("pretty-ms"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const d = (0, _debug).default('electron-forge:async-ora');
const useFakeOra = Boolean(process.env.DEBUG && process.env.DEBUG.includes('electron-forge'));
if (useFakeOra) {
    console.warn(_chalk.default.red('WARNING: DEBUG environment variable detected. Append the electron-forge:lifecycle namespace'));
    console.warn(_chalk.default.red('to the value of DEBUG in order to view progress indicators.'));
}
const fakeOra = (name)=>{
    let oraName = name;
    let startTime = null;
    const timing = ()=>startTime ? `-- after ${_chalk.default.cyan(`${(0, _prettyMs).default(Date.now() - startTime)}`)}` : null
    ;
    const fake = {
        start: ()=>{
            startTime = Date.now();
            d('Process Started:', fake.text);
            return fake;
        },
        fail: ()=>{
            d(_chalk.default.red(`Process Failed: ${fake.text}`), timing());
            return fake;
        },
        succeed: ()=>{
            d('Process Succeeded:', fake.text, timing());
            return fake;
        },
        stop: ()=>{
            d('Process Stopped:', fake.text, timing());
            return fake;
        },
        warn: (warning)=>{
            d('Process Warned:', warning, timing());
            return fake;
        },
        get text () {
            return oraName;
        },
        set text (newName){
            d('Process Renamed:', oraName, ' --> ', newName);
            oraName = newName;
        }
    };
    return fake;
};
exports.fakeOra = fakeOra;
var _default = useFakeOra ? fakeOra : _ora.default;
exports.default = _default;
