// @logseq/capacitor-file-sync/dist/esm/web.js
/* eslint-disable @typescript-eslint/no-unused-vars */
import { WebPlugin } from '@capacitor/core';
export class FileSyncWeb extends WebPlugin {
    cancelAllRequests(_options) {
        throw new Error('Method not implemented.');
    }
    keygen() {
        throw new Error('Method not implemented.');
    }
    setEnv(_options) {
        throw new Error('Method not implemented.');
    }
    encryptFnames(_options) {
        throw new Error('Method not implemented.');
    }
    decryptFnames(_options) {
        throw new Error('Method not implemented.');
    }
    getLocalFilesMeta(_options) {
        throw new Error('Method not implemented.');
    }
    getLocalAllFilesMeta(_options) {
        throw new Error('Method not implemented.');
    }
    deleteLocalFiles(_options) {
        throw new Error('Method not implemented.');
    }
    updateLocalFiles(_options) {
        throw new Error('Method not implemented.');
    }
    updateLocalVersionFiles(_options) {
        throw new Error('Method not implemented.');
    }
    deleteRemoteFiles(_options) {
        throw new Error('Method not implemented.');
    }
    updateRemoteFiles(_options) {
        throw new Error('Method not implemented.');
    }
    encryptWithPassphrase(_options) {
        throw new Error('Method not implemented.');
    }
    decryptWithPassphrase(_options) {
        throw new Error('Method not implemented.');
    }
}
